File: wpimath/src/test/native/cpp/system/struct/GearboxStructTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <gtest/gtest.h>

#include "wpi/math/system/Gearbox.hpp"

using namespace wpi::math;

using StructType = wpi::util::Struct<wpi::math::Gearbox>;

inline constexpr Gearbox kExpectedData =
    Gearbox{DCMotor{1.91_V, 19.1_Nm, 1.74_A, 2.29_A, 2.2_rad_per_s}, 3, 8.45};

TEST(GearboxStructTest, Roundtrip) {
  uint8_t buffer[StructType::GetSize()];
  std::memset(buffer, 0, StructType::GetSize());
  StructType::Pack(buffer, kExpectedData);

  Gearbox unpacked_data = StructType::Unpack(buffer);

  EXPECT_EQ(kExpectedData.numMotors, unpacked_data.numMotors);
  EXPECT_EQ(kExpectedData.reduction, unpacked_data.reduction);
  EXPECT_EQ(kExpectedData.motor.nominalVoltage.value(),
            unpacked_data.motor.nominalVoltage.value());
  EXPECT_EQ(kExpectedData.motor.stallTorque.value(),
            unpacked_data.motor.stallTorque.value());
  EXPECT_EQ(kExpectedData.motor.stallCurrent.value(),
            unpacked_data.motor.stallCurrent.value());
  EXPECT_EQ(kExpectedData.motor.freeCurrent.value(),
            unpacked_data.motor.freeCurrent.value());
  EXPECT_EQ(kExpectedData.motor.freeSpeed.value(),
            unpacked_data.motor.freeSpeed.value());
}

File: wpimath/src/test/java/org/wpilib/math/system/struct/GearboxStructTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.math.system.struct;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import org.junit.jupiter.api.Test;
import org.wpilib.math.system.DCMotor;
import org.wpilib.math.system.Gearbox;

class GearboxStructTest {
  private static final Gearbox DATA =
      new Gearbox(new DCMotor(1.91, 19.1, 1.74, 1.74, 22.9), 3, 8.45);

  @Test
  void testRoundtrip() {
    ByteBuffer buffer = ByteBuffer.allocate(Gearbox.struct.getSize());
    buffer.order(ByteOrder.LITTLE_ENDIAN);
    Gearbox.struct.pack(buffer, DATA);
    buffer.rewind();

    Gearbox data = Gearbox.struct.unpack(buffer);
    assertEquals(DATA.numMotors, data.numMotors);
    assertEquals(DATA.reduction, data.reduction);
    assertEquals(DATA.motor.nominalVoltage, data.motor.nominalVoltage);
    assertEquals(DATA.motor.stallTorque, data.motor.stallTorque);
    assertEquals(DATA.motor.stallCurrent, data.motor.stallCurrent);
    assertEquals(DATA.motor.freeCurrent, data.motor.freeCurrent);
    assertEquals(DATA.motor.freeSpeed, data.motor.freeSpeed);
  }
}
